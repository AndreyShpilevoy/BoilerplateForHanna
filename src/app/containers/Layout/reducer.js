import {getLocaleApi} from 'api';
import {all, call, put, take, fork} from 'redux-saga/effects';

const initialState = {
    locale: 'EN'
};

export const GET_LOCALE = 'GET_LOCALE';
export const getLocale = () => ({type: GET_LOCALE});
export const GET_LOCALE_SUCCESS = 'GET_LOCALE_SUCCESS';
export const getLocaleSuccess = locale => ({type: GET_LOCALE_SUCCESS, payload: {locale} });
export const GET_LOCALE_FAILED = 'GET_LOCALE_FAILED';
export const getLocaleError = error => ({type: GET_LOCALE_FAILED, payload: {error} });

export const layoutReducer = (state = initialState, {type, payload}) => {
    switch (type) {
        case GET_LOCALE_SUCCESS:
            return {...state, locale: payload.locale};
        case GET_LOCALE_FAILED:
            return {...state, error: payload.error};
        default:
            break;
    }
    return state;
};

/* eslint-disable func-style, func-names */
export const getLocaleNonBlockSaga = function* () {
    const {locale, error} = yield call(getLocaleApi);
    if (locale) {
        yield put(getLocaleSuccess(locale));
    } else {
        yield put(getLocaleError(error));
    }
};

export const getLocaleSaga = function* () {
    for (;;) {
        yield take(GET_LOCALE);
        yield fork(getLocaleNonBlockSaga);
    }
};

export const layoutSaga = function* () {
    yield all([
        getLocaleSaga()
    ]);
};
/* eslint-enable func-style, func-names */