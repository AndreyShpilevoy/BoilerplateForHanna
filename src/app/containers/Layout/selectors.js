// import {createSelector} from 'reselect';

export const localeSelector = state => state.layoutReducer.locale; // eslint-disable-line import/prefer-default-export