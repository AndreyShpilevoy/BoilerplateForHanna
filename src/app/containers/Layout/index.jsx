import React, {PureComponent} from 'react';
import {node, func, string} from 'prop-types';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import {getLocale} from './reducer';
import {localeSelector} from './selectors';

export class LayoutPure extends PureComponent {
    static propTypes = {
        children: node.isRequired,
        getLocale: func.isRequired,
        locale: string.isRequired
    };

    componentDidMount() {
        this.props.getLocale();
    }

    render() {
        const localeString = `Locale: ${this.props.locale}`;
        return (
            <div>
                <div>
                    {'Some layout wrapper start'}
                </div>
                <div>
                    {localeString}
                </div>
                {this.props.children}
                <div>
                    {'Some layout wrapper end'}
                </div>
            </div>
        );
    }
}


const mapStateToProps = state => ({
    locale: localeSelector(state)
});

const mapDispatchToProps = dispatch =>
    bindActionCreators({
        getLocale
    }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(LayoutPure);
