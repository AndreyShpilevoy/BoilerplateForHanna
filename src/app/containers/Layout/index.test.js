/* eslint no-undef: 0, react/no-multi-comp:0, react/prop-types:0 react/display-name: 0, react/jsx-filename-extension: 0 */

import React from 'react';
import {mount, shallow, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import configureMockStore from 'redux-mock-store';
import Layout, {LayoutPure} from './index';

configure({adapter: new Adapter()});

jest.mock('./reducer', () => ({
    getLocale: () => ({type: 'MOCK'})
}));

describe('MainPage', () => {
    const mockStore = configureMockStore();

    it('LocalePure match expected snapshot', () => {
        const props = {
            locale: 'EN',
            children: 'content',
            getLocale: () => {}
        };
        expect(shallow(<LayoutPure {...props} />)).toMatchSnapshot();
    });

    it('MainPage match expected snapshot', () => {
        const props = {
            store: mockStore({layoutReducer: {locale: 'EN'} })
        };
        expect(mount(<Layout {...props}>content</Layout>, {lifecycleExperimental: true})).toMatchSnapshot();
    });
});
