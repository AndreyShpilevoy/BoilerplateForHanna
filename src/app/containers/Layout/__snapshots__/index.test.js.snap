// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage LocalePure match expected snapshot 1`] = `
<div>
  <div>
    Some layout wrapper start
  </div>
  <div>
    Locale: EN
  </div>
  content
  <div>
    Some layout wrapper end
  </div>
</div>
`;

exports[`MainPage MainPage match expected snapshot 1`] = `
<Connect(LayoutPure)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <LayoutPure
    getLocale={[Function]}
    locale="EN"
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
  >
    <div>
      <div>
        Some layout wrapper start
      </div>
      <div>
        Locale: EN
      </div>
      content
      <div>
        Some layout wrapper end
      </div>
    </div>
  </LayoutPure>
</Connect(LayoutPure)>
`;
